### PR TITLE
MM-32654 - OpenID Connect Self-Managed

### PR DESCRIPTION
**Pull request by cwarnermm**
Documentation for: https://mattermost.atlassian.net/browse/MM-27688

Updated:
- Self-Managed Admin Guide > Onboard Users > GitLab Single Sign-On
   - Added instructions for setting up a new GitLab OpenID Connect identity provider (E20) only
   - Updated existing instructions for Team Edition and E10

### CODE REVIEW

**Comment by cwarnermm** [comment]
Note to reviewers: Our GitLab SSO documentation intentionally uses a separate approach from our Google Apps SSO and Office 365 SSO product documentation, since GitLab is a service provider offering in all Mattermost editions, including TE, E10, and E20.

@amyblais FYI: This PR work was done on the ``v5.33-documentation`` branch because it did not need to take advantage of shared resources available via the ``master`` branch.

**Review by sbishel — APPROVED** [review]

File: source/deployment/sso-gitlab.rst [hunk 1]
```diff
@@ -1,62 +1,143 @@
...
+  }
+
+3. Save your changes, then restart your Mattermost server. After the server restarts, users must change their sign-in method before they can sign in with GitLab.
+
+(Optional) Step 3: Force users to sign up using SSO only
+--------------------------------------------------------
+
+To force all users to sign-up with SSO only, go to **System Console > Authentication > Email**, then set **Enable sign-in with email** to **false**. 
+
+Alternatively, add this setting to the ``config.json`` file directly by locating the ``EmailSettings`` section, then set ``EnableSignUpWithEmail`` to ``false``. You must save your changes and restart the Mattermost server. Users must change their sign-in method before they can sign in to Mattermost with GitLab.
```
Comment (justinegeffen) [inline source/deployment/sso-gitlab.rst:143]:
```suggestion
Alternatively, add this setting to the ``config.json`` file directly in the ``EmailSettings`` section and set ``EnableSignUpWithEmail`` to ``false``. You must save your changes and restart the Mattermost server. Users must change their sign-in method before they can sign in to Mattermost with GitLab.
```

Comment (justinegeffen) [reply source/deployment/sso-gitlab.rst:143]: Same comment as above re: users and their sign-in methods.

**Comment by cwarnermm** [comment]
Thanks @justinegeffen for the great review! Regarding your question about restarting the server -- the previous documentation was inconsistent in this regard. Only a couple of platforms were addressed, though more are supported. I'd like to see us offer this kind of information in a single way consistently throughout the product documentation, rather than addressed on a page by page basis.

I'm envisioning creating a topic on how to restart a Mattermost server. I've logged https://mattermost.atlassian.net/browse/MM-33843 to work with R&D to determine the steps to restart a Mattermost server across all platforms supported, and add a page with these details. All topics that include instructions to restart the server can then link to these details. 

Thoughts? Concerns?

File: source/deployment/sso-gitlab.rst [hunk 2]
```diff
@@ -1,62 +1,143 @@
...
 
+Step 1: Add an application to your GitLab account
+-------------------------------------------------
 
-Step 1: Add an OAuth application to your GitLab account
--------------------------------------------------------
-
-1. Sign in to your GitLab account and go to ``https://{gitlab-site-name}/profile/applications``. For *{gitlab-site-name}* use the name of your GitLab instance. If you're using GitLab itself as your OAuth provider, use *gitlab.com*.
+1. Sign in to your GitLab account, then go to ``https://{gitlab-site-name}/profile/applications``. For *{gitlab-site-name}* use the name of your GitLab instance. If you're using GitLab itself as your OAuth provider, use ``gitlab.com``.
```
Comment (justinegeffen) [inline source/deployment/sso-gitlab.rst:12]:
```suggestion
1. Sign in to your GitLab account, then go to ``https://{gitlab-site-name}/profile/applications``. For ``{gitlab-site-name}`` use the name of your GitLab instance. If you're using GitLab itself as your OAuth provider, use ``gitlab.com``.
```

File: source/deployment/sso-gitlab.rst [hunk 3]
```diff
@@ -1,62 +1,143 @@
...
 2. Add a new application:
 
-  a. In the **Name** field, type *Mattermost*.
-  b. In the **Redirect URI** field add the following two lines, using your own value for *{mattermost-site-name}*.
+  a. In the **Name** field, enter ``Mattermost``.
+  b. In the **Redirect URI** field, add the following two lines, using your own value for *{mattermost-site-name}*:
```
Comment (justinegeffen) [inline source/deployment/sso-gitlab.rst:16]:
```suggestion
  b. In the **Redirect URI** field, add the following two lines, using your own value for ``{mattermost-site-name}``:
```

File: source/deployment/sso-gitlab.rst [hunk 4]
```diff
@@ -1,62 +1,143 @@
...
+    "ButtonText": "",
+    "ButtonColor": ""
+  }
+
+3. Save your changes, then restart your Mattermost server. After the server restarts, users must change their sign-in method before they can sign in with GitLab.
+
+(Optional) Step 3: Force users to sign up using SSO only
+--------------------------------------------------------
+
+To force all users to sign-up with SSO only, go to **System Console > Authentication > Email**, then set **Enable sign-in with email** to **false**.
```
Comment (justinegeffen) [inline source/deployment/sso-gitlab.rst:141]:
```suggestion
To force all users to sign up with SSO only, go to **System Console > Authentication > Email**, then set **Enable sign-in with email** to **false**. 
```

**Review by justinegeffen — COMMENTED** [review]
I found some other points I wanted to comment on while I was looking for something else. :)

File: source/deployment/sso-gitlab.rst [hunk 5]
```diff
@@ -1,62 +1,143 @@
...
+    "Enable": true,
+    "Secret": "{mattermost-app-secret-from-gitlab}",
+    "Id": "{mattermost-app-application-id-from-gitlab}",
+    "Scope": "",
+    "AuthEndpoint": "https://{gitlab-site-name}/oauth/authorize",
+    "TokenEndpoint": "https://{gitlab-site-name}/oauth/token",
+    "UserApiEndpoint": "https://{gitlab-site-name}/api/v4/user"
+  }
+  
+3. Save your changes, then restart your Mattermost server. After the server restarts, users must change their sign-in method before they can sign in with GitLab.
```
[the remark this answers is not in the document]
Comment (adamjclarkson) [reply source/deployment/sso-gitlab.rst:67]: Is a restart required? Cloud customers can't restart.

**Review by adamjclarkson — APPROVED** [review]

File: source/deployment/sso-gitlab.rst [hunk 6]
```diff
@@ -1,62 +1,143 @@
...
+    "TokenEndpoint": "https://{gitlab-site-name}/oauth/token",
+    "UserApiEndpoint": "https://{gitlab-site-name}/api/v4/user"
+  }
+
+3. Save your changes, then restart your Mattermost server. After the server restarts, users must change their sign-in method before they can sign in with GitLab.
+
+Mattermost Enterprise Edition E20
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Mattermost Enterprise Edition E20 customers can set up a new GitLab OpenID Connect identity provider, or they can convert an existing OAuth 2.0 configuration for GitLab to the OpenID Connect standard. From Mattermost Enterprise Edition v5.33, OAuth 2.0 is being deprecated and replaced by OpenID Connect. Refer to product documentation to `convert your existing OAuth 2.0 service provider configuration for GitLab <https://docs.mattermost.com/deployment/converting-oauth-service-providers-to-openid-connect.html>`__ to the OpenID Connect Standard.
```
Comment (amyblais) [inline source/deployment/sso-gitlab.rst:106]:
```suggestion
Mattermost Enterprise Edition E20 customers can set up a new GitLab OpenID Connect identity provider, or they can convert an existing OAuth 2.0 configuration for GitLab to the OpenID Connect standard. From Mattermost Enterprise Edition v5.33, OAuth 2.0 is being deprecated and replaced by OpenID Connect. Refer to product documentation to `convert your existing OAuth 2.0 service provider configuration for GitLab <https://docs.mattermost.com/deployment/converting-oauth20-service-providers-to-openidconnect.html>`__ to the OpenID Connect Standard. 
```

File: source/deployment/sso-gitlab.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (cwarnermm) [reply source/deployment/sso-gitlab.rst:67]: @adamjclarkson - A restart is required for self-managed customers. The Cloud version of this content does not include a step to restart the server.

**Comment by justinegeffen** [comment]
Thanks for clarifying that, @cwarnermm. That makes perfect sense, and addresses my point perfectly. :)

**Review by justinegeffen — APPROVED** [review]

**Pull request merged**